### PR TITLE
Support CosmWasm in ibc-proto

### PR DESCRIPTION
**Pull request by romac**
Closes: #1913 

## Description

- Add a new `client` feature to gate the `tonic` client code, implies `std`.
- Add a new `json-schema` feature to derive `schemars::JsonSchema` on some proto types, implies `std`.
- Add `#[serde(default)]` to fields that might be omitted by Golang `omitempty` directive.
- Change serialization of byte arrays to Base64 for compatibility with Go.
- Remove distinction between std and no_std for generated code.

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Passing existing no_std checks. 
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by adizere** [comment]
>Remove distinction between std and no_std for generated code.

💘 

Will review soon, but if @lightyear15 is good with the changes we can LGTM.

**Review by soareschen — APPROVED** [review]

File: proto/Cargo.toml [hunk 1]
```diff
@@ -27,11 +27,15 @@ prost-types = { version = "0.9", default-features = false }
...
 version = "=0.23.5"
 default-features = false
 
 [features]
-default = ["std"]
-std = ["tonic", "tonic/codegen", "tonic/transport", "tonic/prost"]
+default = ["std", "client"]
+client = ["std", "tonic", "tonic/codegen", "tonic/transport", "tonic/prost"]
+json-schema = ["std", "schemars"]
+std = ["base64"]
```
Comment (lightyear15) [inline proto/Cargo.toml:41]: Shouldn't this be the opposite?
```
base64 = ["std"]
```

File: proto/src/lib.rs [hunk 2]
```diff
@@ -218,3 +205,19 @@ pub mod ibc {
 pub mod ics23 {
     include_proto!("ics23.rs");
 }
+
+#[cfg(feature = "std")]
```
Comment (lightyear15) [inline proto/src/lib.rs:209]: In my view, there is no use case where this crate does not need to be compatible with Go, so the ``base64`` encoding/decoding should not be optional.

I don't know what use case you guys have for this crate to support ``no_std``, but you can try to explore [``encode_config_slice``](https://docs.rs/base64/latest/base64/fn.encode_config_slice.html) and [``decode_config_slice``](https://docs.rs/base64/latest/base64/fn.decode_config_slice.html) with ``default-features=false`` for [``base64``](https://docs.rs/crate/base64/latest/source/Cargo.toml) crate

File: proto/Cargo.toml (hunk 1 above)
Comment (romac) [reply proto/Cargo.toml:41]: Here `std` is the feature and `base64` the optional crate that is enabled when the `std` feature is enabled. That said, I agree with your comment below (https://github.com/informalsystems/ibc-rs/pull/1914#discussion_r814571968) and will enable the base64 encoding (and therefore Go compat) for both std and no_std.

File: proto/src/lib.rs (hunk 2 above)
Comment (romac) [reply proto/src/lib.rs]: https://github.com/informalsystems/ibc-rs/pull/1914/commits/245c1e761d2d47c25a1327976f775e046119f1bb

**Comment by romac** [comment]
@lightyear15 Does this work for you?

**Review by lightyear15 — APPROVED** [review]
sorry for the delay.
I am fairly confident that it works. I can compile my small cosmwasm contract and run some unit tests using the cosmwasm vm
:+1: 
Thanks

**Pull request merged**